Align draft with POC for multiproof changes (#305)
cfrg/draft-irtf-cfrg-vdaf: draft-irtf-cfrg-vdaf.md

  the nonce and produces a public share, distributed to each of Aggregators,
  and the corresponding sequence of input shares, one for each Aggregator.
  Depending on the VDAF, the input shares may encode additional information

>   used to verify the recovered output shares (e.g., the "proofs shares" in Prio3

Reviewer: "proofs shares" is grammatically odd. Even though we agreed this is the right term to use in the code, in English we should say this differently. I'd also suggesting hinting here that usually there will be one proof, but there could be more. (See suggestion.)
```suggestion
  used to verify the recovered output shares (e.g., the "shares of the of the proof(s)" in Prio3
```

We remark that, taken together, these three functionalities correspond roughly
to the notion of "Affine-aggregatable encodings (AFEs)" from {{CGB17}}.

### Multiple proofs {#multiproofs}

To improve soundness, the prover can construct multiple unique proofs for
its measurement such that the verifier will only accept the measurement once

> all proofs have been thoroughly verified. Notably, several proofs using smaller
> field can offer the same level of soundness as a single proof using large field.

Reviewer: ```suggestion
all proofs have been verified. Notably, several proofs using a smaller
field can offer the same level of soundness as a single proof using a large field.
```

To generate these proofs for a specific measurement, the prover calls
`Flp.prove` multiple times, each time using an independently generated prover
and joint randomness string. The verifier checks each proof independently,
each time with an independently generated query randomness string. It accepts
the measurement only if all the decision algorithm accepts on each proof.

See {{security-multiproof}} below for discussions on choosing the right number
 of proofs.

> 

Reviewer: ```suggestion
```

We remark that, taken together, these three functionalities correspond roughly
to the notion of "Affine-aggregatable encodings (AFEs)" from {{CGB17}}.

> ### Multiple proofs {#multiproofs}

Reviewer: This section should have a forwards reference to {{security-multiproof}}

We remark that, taken together, these three functionalities correspond roughly
to the notion of "Affine-aggregatable encodings (AFEs)" from {{CGB17}}.

> ### Multiple proofs {#multiproofs}

Author: I add this paragraph after we introduce `FLP` in details since we need these details (proof, Flp.prove, field, soundness, etc) to describe the concept of multiproofs better.  What do you think?

all proofs have been thoroughly verified. Notably, several proofs using smaller
field can offer the same level of soundness as a single proof using large field.

To generate these proofs for a specific measurement, the prover can call
`Flp.prove`, using different sets of prover/joint randomnesses. Then,
the shares of these proofs are distributed to the verifier. Subsequently,
 the verifier utilizes `Flp.query` to compute one share of verifier message
 for each proof share, based on the measurement, the proof share, the joint
 randomness, and unique query randomness. Finally, the
verifier runs `Flp.decide` on each verifier message to make a decision. The
validity of the measurement is determined through the logical conjunction
of all Boolean outputs yielded by `Flp.decide`.

## Construction {#prio3-construction}

This section specifies `Prio3`, an implementation of the `Vdaf` interface
({{vdaf}}). It has two generic parameters: an `Flp` ({{flp}}) and a `Xof`
({{xof}}). Moreover, an associated constant, `PROOFS`, can be defined for
 each instance ({{prio3-instantiations}}). `PROOFS`, with a value within
 the range of `[1, 256)`, denotes the number of proofs ({{multiproofs}})

>  and is set to `1` by default, unless explicitly specified. The associated

Author: I hightlight the default value `1` here. Maybe we don't have to repeat it in each instance?

Reviewer: Works for me.

| `USAGE_JOINT_RAND_PART: Unsigned`  | 7     |
{: #prio3-const title="Constants used by Prio3."}

> A list of Prio3 specific parameters, in {{prio3-specific-params}}, are provided by each instantiation {{prio3-instantiations}}.

Author: In POC, `PROOFS` lives in `class Prio3`. So I didn't add this parameter in VDAF or FLP. What do you think?

Reviewer: I agree I don't think it belongs in either table.

I would probably not create a table just for this. The first paragraph of {{prio3-construction}} (i.e., this section) lists `Xof` and `Flp` as "associated parameters"; I'd define `PROOFS` there. Perhaps call it an "associated constant".

1. Encode the Client's measurement for the FLP

> 2. Shard the measurement into a sequence of measurement shares

Reviewer: I think we need a more gentle introduction to the idea of multi-proofs. Here I think we should speak just in terms of a single proof. Then in a short paragraph below, recall the `PROOFS` parameter and say what it does.

          leader_meas_share,
          Prio3.helper_meas_share(j+1, k_helper_meas_shares[j]),
      )

  # Generate the proofs and shard it into proof(s) shares.
  prove_rands = Prio3.prove_rands(k_prove)
  leader_proofs_share = []
  for _ in range(Prio3.PROOFS):
      prove_rand, prove_rands = front(
          Prio3.Flp.PROVE_RAND_LEN, prove_rands)
      leader_proofs_share += Prio3.Flp.prove(meas, prove_rand, [])
  for j in range(Prio3.SHARES-1):
      leader_proofs_share = vec_sub(
          leader_proofs_share,
          Prio3.helper_proofs_share(j+1, k_helper_proofs_shares[j]),
      )

  # Each Aggregator's input share contains its measurement share
  # and proofs share.
  input_shares = []
  input_shares.append((
      leader_meas_share,
      leader_proofs_share,
      None,
  ))
  for j in range(Prio3.SHARES-1):
      input_shares.append((
          k_helper_meas_shares[j],
          k_helper_proofs_shares[j],
          None,
      ))
  return (None, input_shares)
~~~
{: #prio3-shard-without-joint-rand title="Sharding an encoded measurement without joint randomness."}

The steps in this method are as follows:

1. Shard the encoded measurement into shares

> 1. Generate the proofs and shard the proofs into shares

Reviewer: ```suggestion
1. Generate and shard each proof into shares
```

1. Encode each measurement and proofs share into an input share

Notice that only one pair of measurement and proofs shares (called the "leader"
shares above) are vectors of field elements. The other shares (called the
"helper" shares) are represented instead by XOF seeds, which are expanded into
vectors of field elements.

> The methods on `Prio3` for deriving the prover randomnesses, measurement shares,

Reviewer: "randomnesses" might be the right term if "randomness" would be ambiguous in context. Here I don't think it is. (Similar comment below.)

```suggestion
The methods on `Prio3` for deriving the prover randomness, measurement shares,
```

  for _ in range(Prio3.PROOFS):
      prove_rand, prove_rands = front(
          Prio3.Flp.PROVE_RAND_LEN, prove_rands)
      joint_rand, joint_rands = front(
          Prio3.Flp.JOINT_RAND_LEN, joint_rands)
      leader_proofs_share += Prio3.Flp.prove(meas,
                                              prove_rand, joint_rand)
  for j in range(Prio3.SHARES-1):
      leader_proofs_share = vec_sub(
          leader_proofs_share,
          Prio3.helper_proofs_share(j+1, k_helper_proofs_shares[j]),
      )

  # Each Aggregator's input share contains its measurement share,
  # proofs share, and blind. The public share contains the
  # Aggregators' joint randomness parts.
  input_shares = []
  input_shares.append((
      leader_meas_share,
      leader_proofs_share,
      k_leader_blind,
  ))
  for j in range(Prio3.SHARES-1):
      input_shares.append((
          k_helper_meas_shares[j],
          k_helper_proofs_shares[j],
          k_helper_blinds[j],
      ))
  return (k_joint_rand_parts, input_shares)
~~~
{: #prio3-shard-with-joint-rand title="Sharding an encoded measurement with joint randomness."}

The difference between this procedure and previous one is that here we compute

> joint randomnesses `joint_rands`, split into multiple `joint_rand`

Reviewer: ```suggestion
joint randomnesses `joint_rands`, split it into multiple `joint_rand`,
```

and pass each `joint_rand` to the proof generationg algorithm.
(In {{prio3-shard-without-joint-rand}} the joint randomness is the empty
vector, `[]`.) This requires generating an additional value, called the
"blind", that is incorporated into each input share.

The joint randomnesses computation involves the following steps:

1. Compute a "joint randomness part" from each measurement share and blind
1. Compute a "joint randomness seed" from the joint randomness parts

> 1. Compute multiple joint randomnesses from the joint randomness seed

Reviewer: ```suggestion
1. Compute the joint randomness for each proof evaluation from the joint randomness seed
```

This section describes the process of recovering output shares from the input

> shares. The high-level idea is that each Aggregator first queries its

Reviewer: "proofs share" and "verifiers share" are confusing terms in this context. Don't just search and replace; we need to make sure this paragraph is still understandable.

Here and below.

| `Field`          | `Field64` ({{fields}})       |
{: title="Parameters of validity circuit Count."}

| Parameter        | Value                        |
|:-----------------|:-----------------------------|
| `PROOFS`         | `1`                          |

> {: title="Values for Prio3 specific parameters."}

Reviewer: As I mentioned above, I don't think we should create a table for just one parameter. Idea: Perhaps add the operational parameters (I believe just `Xof` and the `Flp`) to this table as well? Here and below.

that every client can contribute to at most one prefix among the ones being
evaluated by the helpers.

> ## Considerations for multiple proofs (`PROOFS`) {#security-multiproof}

Reviewer: ```suggestion
## Choosing the Number of Proofs to Use for Prio3 {#security-multiproof}
```

> ## Considerations for multiple proofs (`PROOFS`) {#security-multiproof}

Reviewer: Add a newline between the title and the first paragraph
```suggestion
## Considerations for multiple proofs (`PROOFS`) {#security-multiproof}

```

> > TODO on how changing `PROOFS` ({{multiproofs}}) impacts security, in particular when we go

Reviewer: * "TODO"s should be imperative, similarly to how we write commit messages. (E.g., TODO Add guidance for ..."
* Add a reference to the open issue.
* Break lines at 80 characters.

time to developing definitions and security proofs.

Thanks to Junye Chen, Henry Corrigan-Gibbs, Armando Faz-Hernández, Simon

> Friedberger, Tim Geoghegan, Albert Liu, Brandon Pitman, Mariana Raykova, Jacob Rothstein,

Reviewer: Paragraphs should wrap at 80 characters (this line is 88 characters long).
